File: web/views/isolation.js
// 权限隔离：固定探针矩阵 + 目录佐证。约束 6（硬编码只看公共检索空间）
// 意味着这个视图不能做成"切换身份浏览私有文档"——后端的
// /api/isolation/probes 签名里本来就没有身份参数，这里也不提供任何
// 身份选择输入框。全部 textContent，绝不 innerHTML。

import { getJSON } from '../lib/api.js';
import { el } from '../lib/dom.js';
import { dataTable } from '../lib/table.js';

export const id = 'isolation';
export const title = '权限隔离';

const STATUS_LABEL = { passed: '通过', failed: '未通过', skipped: '跳过（无演示数据）' };
const STATUS_CLASS = { passed: 'pass', failed: 'fail', skipped: 'dim' };

// 7 个聚合计数——两张转置表（docs 一张、blocks 一张）共用同一份行定义，
// 行 = 计数字段，列 = 4 个身份分支，foreign_* 行非零标红。
const COUNTER_KEYS = [
  'total',
  'public_rows',
  'user_private_rows',
  'tenant_private_rows',
  'empty_owner_rows',
  'foreign_user_rows',
  'foreign_tenant_rows',
];

function stat(label, value, cls) {
  return el(
    'div',
    { class: 'stat' },
    el('div', { class: 'stat-label dim' }, label),
    el('div', { class: `stat-value mono${cls ? ` ${cls}` : ''}` }, value)
  );
}

// 顶部状态条——只对接口已经算好的标志位计数展示，不合成任何论断句
// （约束 7）：数的是 checks[].status / error_branches[].passed /
// rls_status[].relforcerowsecurity / owner_is_superuser，不是写出来的判断。
function statStrip(catalog, probes) {
  const tally = (status) => probes.checks.filter((c) => c.status === status).length;
  const errPassed = probes.error_branches.filter((b) => b.passed).length;
  const forceOff = catalog.rls_status.filter((r) => !r.relforcerowsecurity).length;
  const superOwner = catalog.rls_status.filter((r) => r.owner_is_superuser).length;
  const failed = tally('failed');
  return el(
    'div',
    { class: 'stat-strip' },
    stat('检查·通过', tally('passed'), 'pass'),
    stat('检查·未通过', failed, failed ? 'fail' : 'dim'),
    stat('检查·跳过', tally('skipped'), 'dim'),
    stat(
      '错误分支符合预期',
      `${errPassed}/${probes.error_branches.length}`,
      errPassed === probes.error_branches.length ? 'pass' : 'fail'
    ),
    stat('relforcerowsecurity=false 的表', forceOff, forceOff ? 'fail' : 'pass'),
    stat('owner_is_superuser=true 的表', superOwner, superOwner ? 'fail' : 'pass'),
    stat('demo_seeded', String(probes.demo_seeded), probes.demo_seeded ? 'pass' : 'dim')
  );
}

function branchHeader(b) {
  return el('span', {}, b.label, el('br'), el('span', { class: 'mono dim' }, `${b.tenant ?? '—'}/${b.user ?? '—'}`));
}

function relationMatrix(branches, relation, title) {
  const columns = [
    { label: '计数', cls: 'mono', get: (key) => key },
    ...branches.map((b) => ({
      label: branchHeader(b),
      cls: 'num',
      get: (key) => b[relation][key],
      cellCls: (key, v) => (key.startsWith('foreign') && v > 0 ? 'fail' : undefined),
    })),
  ];
  return el('section', { class: 'card' }, el('h3', {}, title), dataTable({ columns, rows: COUNTER_KEYS }));
}

const RLS_COLUMNS = [
  { label: '表', get: (r) => r.relname },
  { label: 'relrowsecurity', get: (r) => String(r.relrowsecurity), cellCls: (r) => (r.relrowsecurity ? 'pass' : 'fail') },
  {
    label: 'relforcerowsecurity',
    get: (r) => String(r.relforcerowsecurity),
    cellCls: (r) => (r.relforcerowsecurity ? 'pass' : 'fail'),
  },
  { label: 'owner', get: (r) => r.owner },
  {
    label: 'owner_is_superuser',
    get: (r) => String(r.owner_is_superuser),
    cellCls: (r) => (r.owner_is_superuser ? 'fail' : 'pass'),
  },
];

const VIEW_COLUMNS = [
  { label: 'asof 视图', get: (v) => v.relname },
  { label: 'owner', get: (v) => v.owner },
  {
    label: 'owner_is_superuser',
    get: (v) => String(v.owner_is_superuser),
    cellCls: (v) => (v.owner_is_superuser ? 'fail' : 'pass'),
  },
];

const POLICY_COLUMNS = [
  { label: '表', get: (p) => p.tablename },
  { label: 'cmd', get: (p) => p.cmd },
  { label: 'policy', cls: 'mono', get: (p) => p.policyname },
  { label: 'roles', cls: 'mono', get: (p) => p.roles.join(', ') },
  { label: 'qual/with_check', get: (p) => el('pre', { class: 'mono' }, p.qual ?? p.with_check ?? '—') },
];

function identityDl(identity) {
  const pairs = [
    ['current_user', identity.current_user],
    ['session_user', identity.session_user],
    ['current_user_is_superuser', String(identity.current_user_is_superuser)],
    ['asof.doc_block SELECT', String(identity.asof_doc_block_select)],
    ['core.doc_block SELECT', String(identity.core_doc_block_select)],
    ['core.doc_block INSERT', String(identity.core_doc_block_insert)],
    ['core.document SELECT', String(identity.core_document_select)],
    ['quality.quality_metric SELECT', String(identity.quality_metric_select)],
  ];
  const dl = el('dl', { class: 'kv mono' });
  for (const [k, v] of pairs) dl.append(el('dt', {}, k), el('dd', {}, v));
  return dl;
}

// relforcerowsecurity=false 或 owner_is_superuser=true 任一成立，下面探针
// 矩阵的绿全是假的——rls_status 表因此是前置条件，留在可见区，不进
// <details>。其余 SQL 证据（连接身份、asof 视图属主、policy 原文）只是
// 佐证，默认折叠。
function catalogSection(catalog) {
  return el(
    'section',
    { class: 'card span-2' },
    el('h2', {}, '目录佐证（隔离是否真的生效的前置条件）'),
    dataTable({ columns: RLS_COLUMNS, rows: catalog.rls_status }),
    el(
      'details',
      {},
      el('summary', {}, '连接身份 / asof 视图属主 / policy 原文（SQL 证据）'),
      identityDl(catalog.connection_identity),
      dataTable({ columns: VIEW_COLUMNS, rows: catalog.asof_view_owners }),
      dataTable({ columns: POLICY_COLUMNS, rows: catalog.policies, scroll: true })
    )
  );
}

const CHECK_COLUMNS = [
  { label: '检查', cls: 'mono', get: (c) => c.name },
  { label: '结果', get: (c) => STATUS_LABEL[c.status] ?? c.status, cellCls: (c) => STATUS_CLASS[c.status] },
  { label: '说明', cls: 'dim', get: (c) => c.detail },
];

const ERROR_BRANCH_COLUMNS = [
  { label: '分支', cls: 'mono', get: (b) => b.key },
  { label: 'raised', get: (b) => String(b.raised) },
  { label: 'sqlstate', cls: 'mono', get: (b) => b.sqlstate },
  { label: 'message', get: (b) => el('pre', { class: 'mono' }, b.message_head ?? '（未抛出异常）') },
  {
    label: '判定',
    get: (b) => (b.passed ? '符合预期' : '不符合预期'),
    cellCls: (b) => (b.passed ? 'pass' : 'fail'),
  },
];

export async function render(ctx) {
  const [catalog, probes] = await Promise.all([
    getJSON('/api/isolation/catalog', {}, { asOf: ctx.asOf, signal: ctx.signal }),
    getJSON('/api/isolation/probes', {}, { asOf: ctx.asOf, signal: ctx.signal }),
  ]);

  const container = el('div', {});
  container.append(statStrip(catalog, probes));
  container.append(catalogSection(catalog));

  container.append(
    el(
      'section',
      { class: 'card span-2' },
      el('h2', {}, '固定探针矩阵'),
      el('p', { class: 'dim' }, '身份来自服务端硬编码常量，不接受任何请求参数——这里没有身份选择输入框。'),
      el(
        'p',
        { class: probes.demo_seeded ? 'pass' : 'dim' },
        probes.demo_seeded ? '演示数据存在，以下私有相关检查真实生效' : '演示数据不存在（ragdemo db seed-isolation-demo），私有相关检查标"跳过"'
      ),
      el(
        'div',
        { class: 'card-grid' },
        relationMatrix(probes.identity_branches, 'documents', 'documents 可见行数'),
        relationMatrix(probes.identity_branches, 'blocks', 'blocks 可见行数')
      ),
      el('h3', {}, '判定'),
      dataTable({ columns: CHECK_COLUMNS, rows: probes.checks }),
      el('h3', {}, '错误分支（约束 4/5 的运行时告警器）'),
      dataTable({ columns: ERROR_BRANCH_COLUMNS, rows: probes.error_branches })
    )
  );

  return container;
}

File: packages/ragdemo/src/ragdemo/api/queries/isolation.py
"""权限隔离探针：证明隔离机制确实生效，而不是展示私有内容。

约束 6（`owner_user IS NOT NULL` 的块硬编码不可见）与 ADR-0010:45「硬编码
只看公共检索空间」意味着这个模块**不能**做成"切换身份浏览私有文档"——
那正是诊断界面结构上不允许存在的东西。这里改用固定探针矩阵：身份来自
硬编码的 `PROBES` 元组（不接受请求参数），每个分支只返回聚合计数与
布尔判定，一个文本列都不取。看到的是"隔离机制生效与否"，永远看不到
私有内容本身。
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Final

import psycopg

from ragdemo.api.schemas import (
    AsofViewOwner,
    CatalogResponse,
    CheckResult,
    ConnectionIdentity,
    ErrorBranch,
    IdentityBranch,
    PolicyDefinition,
    ProbesResponse,
    RelationCounts,
    RlsStatus,
)
from ragdemo.api.serialize import as_bool, as_int, as_optional_str, as_str, as_str_list
from ragdemo.seed.isolation_demo import DEMO_TENANT, DEMO_USER_A, DEMO_USER_B
from ragdemo_core.db.session import as_of_session


@dataclass(frozen=True)
class Probe:
    key: str
    label: str
    tenant: str | None
    user: str | None


# 固定矩阵，与请求无关——四个分支能展示出来，是因为身份来自这里，
# 不是来自请求参数。带 `&tenant=&user=` 查询参数发这条路由的响应必须
# 逐字节不变（tests/api/test_isolation_api.py 断言这一点）。
PROBES: Final[tuple[Probe, ...]] = (
    Probe("public", "公共空间（无租户、无用户）", None, None),
    Probe("tenant", "租户上下文", DEMO_TENANT, None),
    Probe("user_a", "用户 A 上下文", DEMO_TENANT, DEMO_USER_A),
    Probe("user_b", "用户 B 上下文", DEMO_TENANT, DEMO_USER_B),
)

_COUNT_SQL_TEMPLATE = """
    SELECT count(*) AS total,
           count(*) FILTER (WHERE owner_tenant IS NULL AND owner_user IS NULL) AS public_rows,
           count(*) FILTER (WHERE owner_user IS NOT NULL) AS user_private_rows,
           count(*) FILTER (
               WHERE owner_tenant IS NOT NULL AND owner_user IS NULL
           ) AS tenant_private_rows,
           count(*) FILTER (WHERE owner_user = '') AS empty_owner_rows,
           count(*) FILTER (
               WHERE owner_user IS NOT NULL
                 AND owner_user <> coalesce(current_setting('app.user', true), '')
           ) AS foreign_user_rows,
           count(*) FILTER (
               WHERE owner_tenant IS NOT NULL
                 AND owner_tenant <> coalesce(current_setting('app.tenant', true), '')
           ) AS foreign_tenant_rows
      FROM {relation}
"""


def _row_to_counts(row: tuple[object, ...]) -> RelationCounts:
    return RelationCounts(
        total=as_int(row[0]),
        public_rows=as_int(row[1]),
        user_private_rows=as_int(row[2]),
        tenant_private_rows=as_int(row[3]),
        empty_owner_rows=as_int(row[4]),
        foreign_user_rows=as_int(row[5]),
        foreign_tenant_rows=as_int(row[6]),
    )


def _run_identity_branch(
    conn: psycopg.Connection[tuple[object, ...]], probe: Probe, as_of: datetime
) -> IdentityBranch:
    """一个独立的顶层事务：`as_of_session` 内部用 `conn.transaction()`
    开启，退出即提交/回滚，不与其余五个分支共享事务边界。"""
    with as_of_session(conn, as_of, tenant=probe.tenant, user=probe.user):
        doc_row = conn.execute(_COUNT_SQL_TEMPLATE.format(relation="asof.document")).fetchone()
        block_row = conn.execute(_COUNT_SQL_TEMPLATE.format(relation="asof.doc_block")).fetchone()
    assert doc_row is not None
    assert block_row is not None
    return IdentityBranch(
        key=probe.key,
        label=probe.label,
        tenant=probe.tenant,
        user=probe.user,
        documents=_row_to_counts(doc_row),
        blocks=_row_to_counts(block_row),
    )


def _missing_as_of_branch(conn: psycopg.Connection[tuple[object, ...]]) -> ErrorBranch:
    """不设 as_of 直接查 asof 视图——期望 22023（约束 4 的运行时告警器）。

    显式 `set_config('app.as_of', '', true)` 而不是"什么都不做"：连接是
    autocommit，每条语句各自成事务，GUC 本来就没设；写空串同时覆盖
    "没设"与"设成空串"两种情况，正对上 `asof.current_as_of()` 里
    `v IS NULL OR v = ''` 的双判。异常在 `conn.transaction()` 块内抛出，
    退出时自动 ROLLBACK，连接随即恢复可用——不能把 try/except 包在事务块
    外面再手动处理，那样异常发生时事务不会正确回滚。
    """
    try:
        with conn.transaction():
            conn.execute("SELECT set_config('app.as_of', '', true)")
            conn.execute("SELECT count(*) FROM asof.doc_block")
    except psycopg.errors.InvalidParameterValue as exc:
        return ErrorBranch(
            key="missing_as_of",
            raised=True,
            sqlstate=exc.sqlstate,
            message_head=str(exc).splitlines()[0],
            passed=exc.sqlstate == "22023",
        )
    return ErrorBranch(
        key="missing_as_of", raised=False, sqlstate=None, message_head=None, passed=False
    )


def _base_table_branch(conn: psycopg.Connection[tuple[object, ...]]) -> ErrorBranch:
    """误查 core.* 基表——期望 42501（约束 5 的运行时告警器）。

    在超级用户误配置（开发模式忘了 SET ROLE）下这个分支会**成功**，
    于是 passed=false 亮红——这正是要的运行时体检：约束 5 的机械保证
    只在 app_diag 身份下成立，超级用户连接下它本来就不该假装安全。
    """
    try:
        with conn.transaction():
            conn.execute("SELECT count(*) FROM core.doc_block")
    except psycopg.errors.InsufficientPrivilege as exc:
        return ErrorBranch(
            key="base_table_denied",
            raised=True,
            sqlstate=exc.sqlstate,
            message_head=str(exc).splitlines()[0],
            passed=exc.sqlstate == "42501",
        )
    return ErrorBranch(
        key="base_table_denied", raised=False, sqlstate=None, message_head=None, passed=False
    )


def _status(condition: bool) -> str:
    return "passed" if condition else "failed"


def _compute_checks(branches: list[IdentityBranch]) -> tuple[list[CheckResult], bool]:
    by_key = {b.key: b for b in branches}
    public = by_key["public"]
    tenant = by_key.get("tenant")
    user_a = by_key.get("user_a")

    # demo_seeded 是状态位，不是断言——它只回答"这次演示数据存不存在"，
    # 后面几条私有相关的检查据此判定要不要标 skipped：没有私有数据时
    # foreign_user_rows 之类的计数恒为 0，"passed" 会是一句没有测试过
    # 任何东西的假绿；标 skipped 更诚实。
    demo_seeded = user_a is not None and user_a.documents.user_private_rows > 0

    no_foreign_user = all(
        b.documents.foreign_user_rows == 0 and b.blocks.foreign_user_rows == 0 for b in branches
    )
    no_foreign_tenant = all(
        b.documents.foreign_tenant_rows == 0 and b.blocks.foreign_tenant_rows == 0 for b in branches
    )
    no_empty_owner = all(
        b.documents.empty_owner_rows == 0 and b.blocks.empty_owner_rows == 0 for b in branches
    )
    public_sees_no_private = (
        public.documents.user_private_rows == 0
        and public.documents.tenant_private_rows == 0
        and public.blocks.user_private_rows == 0
        and public.blocks.tenant_private_rows == 0
    )
    tenant_sees_no_user_private = tenant is None or (
        tenant.documents.user_private_rows == 0 and tenant.blocks.user_private_rows == 0
    )
    monotonic = all(
        b.documents.total >= public.documents.total and b.blocks.total >= public.blocks.total
        for b in branches
    )
    not_silently_empty = public.documents.total > 0

    checks = [
        CheckResult(
            name="demo_seeded", status=_status(demo_seeded), detail=f"演示数据存在: {demo_seeded}"
        ),
        CheckResult(
            name="no_foreign_user_row",
            status=_status(no_foreign_user) if demo_seeded else "skipped",
            detail="任何分支看见的私有行，owner_user 必须恰好是本分支身份",
        ),
        CheckResult(
            name="no_foreign_tenant_row",
            status=_status(no_foreign_tenant) if demo_seeded else "skipped",
            detail="任何分支看见的租户私有行，owner_tenant 必须恰好是本分支租户",
        ),
        CheckResult(
            name="public_sees_no_private",
            status=_status(public_sees_no_private) if demo_seeded else "skipped",
            detail="公共分支不应看见任何私有/租户行",
        ),
        CheckResult(
            name="tenant_sees_no_user_private",
            status=_status(tenant_sees_no_user_private) if demo_seeded else "skipped",
            detail="租户分支不应看见用户私有行",
        ),
        CheckResult(
            name="owner_user_never_empty_string",
            status=_status(no_empty_owner) if demo_seeded else "skipped",
            detail="owner_user/owner_tenant 为空字符串会匹配 RLS 的默认值，等于向所有人开放",
        ),
        CheckResult(
            name="monotonic_vs_public",
            status=_status(monotonic),
            detail="加身份只能让可见行数增加，不能减少",
        ),
        CheckResult(
            name="not_silently_empty",
            status=_status(not_silently_empty),
            detail="公共分支总行数必须大于 0，防止「全都看不见」被误读成「隔离生效」",
        ),
    ]
    return checks, demo_seeded


def run_probes(conn: psycopg.Connection[tuple[object, ...]], as_of: datetime) -> ProbesResponse:
    """六个各自独立的顶层事务，顺序固定：先四个读分支，再两个故意报错的
    分支——这样即使某天错误分支把连接搞坏了，前四个分支的数据也已经取到。
    """
    branches = [_run_identity_branch(conn, p, as_of) for p in PROBES]
    error_branches = [_missing_as_of_branch(conn), _base_table_branch(conn)]
    checks, demo_seeded = _compute_checks(branches)
    return ProbesResponse(
        as_of=as_of.isoformat(),
        demo_seeded=demo_seeded,
        identity_branches=branches,
        checks=checks,
        error_branches=error_branches,
    )


_POLICIES_SQL = """
    SELECT schemaname, tablename, policyname, cmd, roles::text[], qual, with_check
      FROM pg_policies WHERE schemaname = 'core' AND tablename IN ('document', 'doc_block')
     ORDER BY tablename, cmd, policyname
"""

_RLS_STATUS_SQL = """
    SELECT c.relname, c.relrowsecurity, c.relforcerowsecurity, r.rolname, r.rolsuper
      FROM pg_class c JOIN pg_roles r ON r.oid = c.relowner
     WHERE c.relnamespace = 'core'::regnamespace AND c.relname IN ('document', 'doc_block')
     ORDER BY c.relname
"""

_ASOF_VIEW_OWNERS_SQL = """
    SELECT c.relname, r.rolname, r.rolsuper
      FROM pg_class c JOIN pg_roles r ON r.oid = c.relowner
     WHERE c.relnamespace = 'asof'::regnamespace AND c.relkind = 'v'
     ORDER BY c.relname
"""

_CONNECTION_IDENTITY_SQL = """
    SELECT current_user, session_user,
           coalesce((SELECT rolsuper FROM pg_roles WHERE rolname = current_user), false),
           has_table_privilege(current_user, 'asof.doc_block', 'SELECT'),
           has_table_privilege(current_user, 'core.doc_block', 'SELECT'),
           has_table_privilege(current_user, 'core.doc_block', 'INSERT'),
           has_table_privilege(current_user, 'core.document', 'SELECT'),
           has_table_privilege(current_user, 'quality.quality_metric', 'SELECT')
"""


def get_catalog(
    conn: psycopg.Connection[tuple[object, ...]], *, as_of: datetime
) -> CatalogResponse:
    """隔离机制是否真在生效的目录佐证——原样回传 `pg_policies` 的策略原文
    与 `relforcerowsecurity`/属主超级用户判定，不加解读。这不是探针矩阵的
    补充说明，是它的前置条件：`relforcerowsecurity=false` 或
    `owner_is_superuser=true` 任一成立，前面四个身份分支的绿全是假的。
    """
    policy_rows = conn.execute(_POLICIES_SQL).fetchall()
    rls_rows = conn.execute(_RLS_STATUS_SQL).fetchall()
    view_rows = conn.execute(_ASOF_VIEW_OWNERS_SQL).fetchall()
    identity_row = conn.execute(_CONNECTION_IDENTITY_SQL).fetchone()
    assert identity_row is not None

    return CatalogResponse(
        as_of=as_of.isoformat(),
        # 这条查询完全不看 as_of——目录信息是数据库结构状态，不随时点变化；
        # 参数仍然必填（约束 4 不留例外），这个字段如实标注它对结果无影响。
        as_of_affects_result=False,
        policies=[
            PolicyDefinition(
                schemaname=as_str(r[0]),
                tablename=as_str(r[1]),
                policyname=as_str(r[2]),
                cmd=as_str(r[3]),
                roles=as_str_list(r[4]),
                qual=as_optional_str(r[5]),
                with_check=as_optional_str(r[6]),
            )
            for r in policy_rows
        ],
        rls_status=[
            RlsStatus(
                relname=as_str(r[0]),
                relrowsecurity=as_bool(r[1]),
                relforcerowsecurity=as_bool(r[2]),
                owner=as_str(r[3]),
                owner_is_superuser=as_bool(r[4]),
            )
            for r in rls_rows
        ],
        asof_view_owners=[
            AsofViewOwner(
                relname=as_str(r[0]), owner=as_str(r[1]), owner_is_superuser=as_bool(r[2])
            )
            for r in view_rows
        ],
        connection_identity=ConnectionIdentity(
            current_user=as_str(identity_row[0]),
            session_user=as_str(identity_row[1]),
            current_user_is_superuser=as_bool(identity_row[2]),
            asof_doc_block_select=as_bool(identity_row[3]),
            core_doc_block_select=as_bool(identity_row[4]),
            core_doc_block_insert=as_bool(identity_row[5]),
            core_document_select=as_bool(identity_row[6]),
            quality_metric_select=as_bool(identity_row[7]),
        ),
    )


__all__ = ("PROBES", "Probe", "get_catalog", "run_probes")
